File: my-folder/0200-number-of-islands/solution.java
class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }
        
        int rows = grid.length;
        int cols = grid[0].length;
        int count = 0;
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    dfs(grid, j, i);
                }
            }
        }
        return count;
    }

    void dfs(char[][] grid, int x, int y) {
        if (x < 0 || x >= grid[0].length || y < 0 || y >= grid.length || grid[y][x] != '1') {
            return;
        }
        grid[y][x] = '0';
        dfs(grid, x+1, y);
        dfs(grid, x-1, y);
        dfs(grid, x, y+1);
        dfs(grid, x, y-1);
    }
}

File: my-folder/0200-number-of-islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(x, y, grid):
            if x < 0 or x >= len(grid[0]) or y < 0 or y >= len(grid) or grid[y][x] == "0":
                return
            grid[y][x] = "0"
            dfs(x+1, y, grid)
            dfs(x-1, y, grid)
            dfs(x, y+1, grid)
            dfs(x, y-1, grid)
        
        count = 0

        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == "1":
                    count += 1
                    dfs(c, r, grid)
        
        return count
